### ISSUE TITLE
Unable to build (what compiler should be used?)

### ISSUE THREAD

**Issue by SolidHal** [lkorigin/laniakea#2]
I've started assembling some build instructions that I'll submit as a pull request when complete but I'm running into some issues. 

First, heres what I've found gets me the farthest:

make a staging directory `mkdir laniakea_stage && cd laniakea_stage`
Clone https://github.com/lkorigin/laniakea
install meson `pip3 install meson` 
add meson to path: `export PATH=$PATH:~/.local/bin`
install prereqs: `apt install pkg-config libarchive-dev libcurl4-openssl-dev libczmq-dev libglib2.0-dev appstream libappstream-dev libpq-dev vibe.d  gir-to-d  libstdx-allocator-dev libdcontainers-dev`

Now to choose a compiler from `dmd` ,`gdc`, or `ldc`
`gdc` and `ldc` are both available in debian as packages and dmd can be found as a .deb on their website.

run meson: `meson laniakea`
run ninja `ninja install`

All three compilers throw different errors

***TL;DR*** Which compiler _should_  work? 

`dmd` throws:

```
[165/323] Compiling D object 'src/laniakea/src@laniakea@@laniakea@sta/net.d.o'.
FAILED: src/laniakea/src@laniakea@@laniakea@sta/net.d.o 
dmd -Isrc/laniakea/src@laniakea@@laniakea@sta -Isrc/laniakea -Ilaniakea/src/laniakea -Isrc -Ilaniakea/src -Ilaniakea/contrib/subprojects/dyaml/source/ -Ilaniakea/contrib/subprojects/tinyendian/source/ -Ilaniakea/contrib/subprojects/ddbc/source/ -Ilaniakea/contrib/subprojects/derelict-pq/source/ -Ilaniakea/contrib/subprojects/hibernated/source/ -Icontrib/wrap/../../wrap -Ilaniakea/contrib/wrap/../../wrap -I/usr/include/d/containers -I/usr/include/d/stdx-allocator -I/usr/include/x86_64-linux-gnu -I/usr/include/d/vibe -I/usr/include/d/glibd-2 -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -I/usr/include/appstream -color=on -wi -g -version=Derelict_Static -version=USE_PGSQL -version=Derelict_Static -version=USE_PGSQL -version=Derelict_Static -version=Derelict_Static -version=USE_PGSQL -version=Derelict_Static -J/home/user/Linux_Builds/laniakea-build/laniakea/data/   -of'src/laniakea/src@laniakea@@laniakea@sta/net.d.o' -c laniakea/src/laniakea/net.d
laniakea/src/laniakea/net.d(127): Error: module `stdio` is in file 'core/sys/linux/stdio.d' which cannot be read
```

`gdc` throws 
```
[12/323] Compiling D object 'contrib/subp...yaml@@dyaml@sha/source_dyaml_loader.d.o'.
FAILED: contrib/subprojects/dyaml/contrib@subprojects@dyaml@@dyaml@sha/source_dyaml_loader.d.o 
gdc -Icontrib/subprojects/dyaml/contrib@subprojects@dyaml@@dyaml@sha -Icontrib/subprojects/dyaml -Ilaniakea/contrib/subprojects/dyaml -Ilaniakea/contrib/subprojects/dyaml/source/ -Ilaniakea/contrib/subprojects/tinyendian/source/ -fdiagnostics-color=always -Wall -Wdeprecated -g -O0 -fPIC  -MD -MQ 'contrib/subprojects/dyaml/contrib@subprojects@dyaml@@dyaml@sha/source_dyaml_loader.d.o' -MF 'contrib/subprojects/dyaml/contrib@subprojects@dyaml@@dyaml@sha/source_dyaml_loader.d.o.deps' -o 'contrib/subprojects/dyaml/contrib@subprojects@dyaml@@dyaml@sha/source_dyaml_loader.d.o' -c laniakea/contrib/subprojects/dyaml/source/dyaml/loader.d
laniakea/contrib/subprojects/dyaml/source/dyaml/loader.d:230:31: error: this.opApply matches more than one declaration:
   laniakea/contrib/subprojects/dyaml/source/dyaml/loader.d:245:13:     @safe int(int delegate(ref Node) @safe dg)
and:
   laniakea/contrib/subprojects/dyaml/source/dyaml/loader.d:250:13:     @system int(int delegate(ref Node) @system dg)
             foreach(ref node; this)
                               ^
laniakea/contrib/subprojects/dyaml/source/dyaml/loader.d:230:13: error: cannot uniquely infer foreach argument types
             foreach(ref node; this)
             ^
[14/323] Compiling D object 'contrib/subp...@dyaml@sha/source_dyaml_constructor.d.o'.
ninja: build stopped: subcommand failed.

```

and `ldc2` throws 

```
[301/323] Compiling D object 'src/web/src@web@@laniakea-web@exe/app.d.o'.
FAILED: src/web/src@web@@laniakea-web@exe/app.d.o 
ldc2 -Isrc/web/src@web@@laniakea-web@exe -Isrc/web -Ilaniakea/src/web -Isrc -Ilaniakea/src -Icontrib/wrap/../../wrap -Ilaniakea/contrib/wrap/../../wrap -Ilaniakea/contrib/subprojects/dyaml/source/ -Ilaniakea/contrib/subprojects/tinyendian/source/ -Ilaniakea/contrib/subprojects/ddbc/source/ -Ilaniakea/contrib/subprojects/derelict-pq/source/ -Ilaniakea/contrib/subprojects/hibernated/source/ -I/usr/include/d/containers -I/usr/include/d/stdx-allocator -I/usr/include/x86_64-linux-gnu -I/usr/include/d/vibe -I/usr/include/d/glibd-2 -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -I/usr/include/appstream -I/usr/include/d/diet -enable-color -wi -g -O0 -d-version=Derelict_Static -d-version=USE_PGSQL -d-version=Derelict_Static -d-version=USE_PGSQL -d-version=Derelict_Static -d-version=Derelict_Static -d-version=USE_PGSQL -d-version=Derelict_Static -J/home/user/Linux_Builds/laniakea-build/laniakea/src/web/views/ -d-version=VibeDefaultMain -d-version=Have_diet_ng   -of 'src/web/src@web@@laniakea-web@exe/app.d.o' -c laniakea/src/web/app.d
/usr/include/d/vibe/vibe/core/log.d(565): Deprecation: Built-in hex string literals are deprecated, use `std.conv.hexString` instead.
/usr/include/d/vibe/vibe/http/common.d(488): Error: cannot implicitly convert expression `theAllocator()` of type `RCIAllocator` to `std.experimental.allocator.IAllocator`
/usr/include/d/diet/diet/input.d(7): Deprecation: module `diet.traits` member `DietTraitsAttribute` is not visible from module `input`
laniakea/src/laniakea/localconfig.d(26): Deprecation: Symbol `std.file.getcwd` is not visible from module `localconfig` because it is privately imported in module `path`
[302/323] Compiling D object 'src/web/src@web@@laniakea-web@exe/webconfig.d.o'.
laniakea/src/laniakea/localconfig.d(26): Deprecation: Symbol `std.file.getcwd` is not visible from module `localconfig` because it is privately imported in module `path`
laniakea/contrib/subprojects/hibernated/source/hibernated/metadata.d-mixin-3358(6397): Deprecation: function `std.exception.enforceEx!(MappingException).enforceEx!bool.enforceEx` is deprecated - Use `enforce`. `enforceEx` will be removed with 2.089.
laniakea/contrib/subprojects/hibernated/source/hibernated/metadata.d-mixin-3358(6402): Deprecation: function `std.exception.enforceEx!(MappingException).enforceEx!bool.enforceEx` is deprecated - Use `enforce`. `enforceEx` will be removed with 2.089.
laniakea/contrib/subprojects/hibernated/source/hibernated/metadata.d(3376): Deprecation: function `std.exception.enforceEx!(MappingException).enforceEx!bool.enforceEx` is deprecated - Use `enforce`. `enforceEx` will be removed with 2.089.
laniakea/contrib/subprojects/hibernated/source/hibernated/metadata.d(3381): Deprecation: function `std.exception.enforceEx!(MappingException).enforceEx!bool.enforceEx` is deprecated - Use `enforce`. `enforceEx` will be removed with 2.089.
laniakea/contrib/subprojects/hibernated/source/hibernated/metadata.d(3386): Deprecation: function `std.exception.enforceEx!(MappingException).enforceEx!bool.enforceEx` is deprecated - Use `enforce`. `enforceEx` will be removed with 2.089.
laniakea/contrib/subprojects/hibernated/source/hibernated/metadata.d(3391): Deprecation: function `std.exception.enforceEx!(MappingException).enforceEx!bool.enforceEx` is deprecated - Use `enforce`. `enforceEx` will be removed with 2.089.
laniakea/contrib/subprojects/hibernated/source/hibernated/metadata.d(3396): Deprecation: function `std.exception.enforceEx!(MappingException).enforceEx!bool.enforceEx` is deprecated - Use `enforce`. `enforceEx` will be removed with 2.089.
[303/323] Linking target src/ariadne/lk_test_self-ariadne.
/usr/bin/ld: warning: libphobos2-ldc-shared.so.78, needed by /usr/lib/gcc/x86_64-linux-gnu/6/../../../x86_64-linux-gnu/libdcontainers.so, may conflict with libphobos2-ldc-shared.so.81
/usr/bin/ld: warning: libdruntime-ldc-shared.so.78, needed by /usr/lib/gcc/x86_64-linux-gnu/6/../../../x86_64-linux-gnu/libdcontainers.so, may conflict with libdruntime-ldc-shared.so.81
/usr/bin/ld: warning: libssl.so.1.0.2, needed by /usr/lib/gcc/x86_64-linux-gnu/6/../../../x86_64-linux-gnu/libcurl.so, may conflict with libssl.so.1.1
/usr/bin/ld: warning: libssl.so.1.0.2, needed by /usr/lib/gcc/x86_64-linux-gnu/6/../../../x86_64-linux-gnu/libcurl.so, may conflict with libssl.so.1.1
/usr/bin/ld: warning: libcrypto.so.1.0.2, needed by /usr/lib/gcc/x86_64-linux-gnu/6/../../../x86_64-linux-gnu/libcurl.so, may conflict with libcrypto.so.1.1
/usr/bin/ld: warning: libcrypto.so.1.0.2, needed by /usr/lib/gcc/x86_64-linux-gnu/6/../../../x86_64-linux-gnu/libcurl.so, may conflict with libcrypto.so.1.1
ninja: build stopped: subcommand failed.

```

If it's helpful, the meson output is:

```
The Meson build system
Version: 0.47.2
Source dir: /home/user/Linux_Builds/laniakea-build/laniakea
Build dir: /home/user/Linux_Builds/laniakea-build
Build type: native build
Project name: Laniakea
Project version: 0.1
Native D compiler: ldc2 (llvm 1.11.0 "LDC - the LLVM D compiler (1.11.0):")
Build machine cpu family: x86_64
Build machine cpu: x86_64
Found pkg-config: /usr/bin/pkg-config (0.29)
Native dependency libcrypto found: YES 1.1.0f
Native dependency libssl found: YES 1.1.0f
Native dependency libarchive found: YES 3.2.2
Native dependency libcurl found: YES 7.52.1
Native dependency libczmq found: YES 4.0.2
Native dependency systemd found: YES 232
Native dependency dcontainers found: YES 0.8.0
Looking for a fallback subproject for the dependency dyaml
Cloning into 'dyaml'...
remote: Counting objects: 5361, done.
remote: Compressing objects: 100% (4/4), done.
remote: Total 5361 (delta 1), reused 3 (delta 1), pack-reused 5356
Receiving objects: 100% (5361/5361), 3.57 MiB | 2.51 MiB/s, done.
Resolving deltas: 100% (3048/3048), done.

|
|Executing subproject dyaml 
|
|Project name: D-YAML
|Project version: 0.7.0
|Native D compiler: ldc2 (llvm 1.11.0 "LDC - the LLVM D compiler (1.11.0):")
|Looking for a fallback subproject for the dependency tinyendian
Cloning into 'tinyendian'...
remote: Counting objects: 102, done.
remote: Total 102 (delta 0), reused 0 (delta 0), pack-reused 102
Receiving objects: 100% (102/102), 38.47 KiB | 0 bytes/s, done.
Resolving deltas: 100% (43/43), done.
|
||
||Executing subproject tinyendian 
||
||Project name: tinyendian
||Project version: 0.2.0
||Native D compiler: ldc2 (llvm 1.11.0 "LDC - the LLVM D compiler (1.11.0):")
||Build targets in project: 1
||
||Subproject tinyendian finished.
|Dependency tinyendian from subproject contrib/subprojects/tinyendian found: YES
|Build targets in project: 2
|
|Subproject dyaml finished.
Dependency dyaml from subproject contrib/subprojects/dyaml found: YES
Looking for a fallback subproject for the dependency fluent-asserts
Cloning into 'fluent-asserts'...
remote: Counting objects: 1008, done.
remote: Total 1008 (delta 0), reused 0 (delta 0), pack-reused 1008
Receiving objects: 100% (1008/1008), 136.24 KiB | 0 bytes/s, done.
Resolving deltas: 100% (503/503), done.

|
|Executing subproject fluent-asserts 
|
|Project name: fluent-asserts
|Project version: 0.6.7
|Native D compiler: ldc2 (llvm 1.11.0 "LDC - the LLVM D compiler (1.11.0):")
|Looking for a fallback subproject for the dependency ddmp
Cloning into 'ddmp'...
remote: Counting objects: 161, done.
remote: Total 161 (delta 0), reused 0 (delta 0), pack-reused 161
Receiving objects: 100% (161/161), 65.26 KiB | 0 bytes/s, done.
Resolving deltas: 100% (59/59), done.
|
||
||Executing subproject ddmp 
||
||Project name: ddmp
||Project version: 0.0.1
||Native D compiler: ldc2 (llvm 1.11.0 "LDC - the LLVM D compiler (1.11.0):")
||Build targets in project: 4
||
||Subproject ddmp finished.
|Dependency ddmp from subproject contrib/subprojects/ddmp found: YES
|Build targets in project: 6
|
|Subproject fluent-asserts finished.
Dependency fluent-asserts from subproject contrib/subprojects/fluent-asserts found: YES
Native dependency glibd-2.0 found: YES 2.0.0
Native dependency appstream found: YES 0.12.2
Looking for a fallback subproject for the dependency ddbc-pgsql
Cloning into 'ddbc'...
remote: Counting objects: 1968, done.
remote: Total 1968 (delta 0), reused 0 (delta 0), pack-reused 1968
Receiving objects: 100% (1968/1968), 5.93 MiB | 3.32 MiB/s, done.
Resolving deltas: 100% (1009/1009), done.

|
|Executing subproject ddbc 
|
|Project name: ddbc
|Project version: 0.3.7
|Native D compiler: ldc2 (llvm 1.11.0 "LDC - the LLVM D compiler (1.11.0):")
|Looking for a fallback subproject for the dependency derelict-pq
Cloning into 'derelict-pq'...
remote: Counting objects: 223, done.
remote: Total 223 (delta 0), reused 0 (delta 0), pack-reused 223
Receiving objects: 100% (223/223), 58.08 KiB | 0 bytes/s, done.
Resolving deltas: 100% (74/74), done.
|
||
||Executing subproject derelict-pq 
||
||Project name: DerelictPQ
||Project version: 3.0.0
||Native D compiler: ldc2 (llvm 1.11.0 "LDC - the LLVM D compiler (1.11.0):")
||Native dependency libpq found: YES 10.5
||Build targets in project: 7
||
||Subproject derelict-pq finished.
|Dependency derelict-pq from subproject contrib/subprojects/derelict-pq found: YES
|Build targets in project: 8
|
|Subproject ddbc finished.
Dependency ddbc-pgsql from subproject contrib/subprojects/ddbc found: YES
Looking for a fallback subproject for the dependency hibernated-pgsql
Cloning into 'hibernated'...
remote: Counting objects: 1592, done.
remote: Total 1592 (delta 0), reused 0 (delta 0), pack-reused 1592
Receiving objects: 100% (1592/1592), 1.04 MiB | 1.84 MiB/s, done.
Resolving deltas: 100% (798/798), done.

|
|Executing subproject hibernated 
|
|Project name: hibernated
|Project version: 0.3.2
|Native D compiler: ldc2 (llvm 1.11.0 "LDC - the LLVM D compiler (1.11.0):")
|Dependency ddbc-pgsql from subproject contrib/subprojects/ddbc found: YES (cached)
|Build targets in project: 9
|
|Subproject hibernated finished.
Dependency hibernated-pgsql from subproject contrib/subprojects/hibernated found: YES
Native dependency vibe-core found: YES 0.8.2
Native dependency vibe-data found: YES 0.8.2
Program girtod found: YES (/usr/bin/girtod)
Message: Generating D interfaces from GIR...
Native dependency vibe-http found: YES 0.8.2
Native dependency vibe-web found: YES 0.8.2
Dependency vibe-http found: YES (cached)
Dependency vibe-web found: YES (cached)
Configuring laniakea-lighthouse.service using configuration
Configuring laniakea-web.service using configuration
Configuring laniakea-webswview.service using configuration
Build targets in project: 35
Found ninja-1.7.2 at /usr/bin/ninja
```

**Comment by ximion** [comment]
Hi!
First of all, the compiler that should be used is whatever the `default-d-compiler` package points to for your platform. So, in this case it's `ldc`.

The issues you are seeing are due to Debian being right in the middle of a LDC transition, as well as some rework going on in Laniakea.
LK is pretty much an internally used project at time, so you might occasionally see breakage (although this should get less and less over time). I am currently working on preparation of a few more invasive changes, after which the time of big refactoring should be over for a while, and the modules can be packaged properly for Debian.

In any case, I'll gladly answer any questions you have, PRs and bug reports are also always welcome!
I'll come back to this bug report when the LDC transition has moved a bit further in Debian testing and all deprecation issues have been dealt with.

**Comment by ximion** [comment]
@SolidHal The LDC transition is now far enough in Debian unstable that building this project is possible again.
Give it a try! By the way, you could just have installed Meson from Debian's repositories, without using pip (`apt install meson`).
Also take a look at https://github.com/lkorigin/laniakea/blob/master/meson_options.txt to see which modules you want to build (they can all be disabled individually).

**Comment by ximion** [comment]
LDC is updated and the transition in Debian is almost complete. Also, the CI builds properly again, so I think this is safe to close.
